Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,299 @@
+            self.stateMachineLock.withLockedValue { $0.sequenceSource?.finish() }
+            continuation?.resume()
+        case .none:
+            break
+        }
+    }
+}
+
+extension KafkaPollingSystem: NIOAsyncSequenceProducerDelegate {
+    func produceMore() {
+        let command = self.stateMachineLock.withLockedValue { $0.produceMore() }
+        self.handleStateMachineCommand(command)
+    }
+
+    func didTerminate() {
+        let command = self.stateMachineLock.withLockedValue { $0.shutDown() }
+        self.handleStateMachineCommand(command)
+    }
+}
+
+extension KafkaPollingSystem {
+    /// The state machine used by the ``KafkaBackPressurePollingSystem``.
+    struct StateMachine {
+        /// A flag that determines if the ``run()`` method has already been invoked.
+        var running = false
+        /// Closure that takes care of polling `librdkafka` for new messages.
+        var pollClosure: (() -> Void)?
+        /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+        var sequenceSource: Producer.Source?
+
+        /// The possible states of the state machine.
+        enum State {
+            /// Initial state.
+            case initial
+            /// The system up and producing acknowledgement messages.
+            case producing
+            /// The pool loop is currently suspended and we are waiting for an invocation
+            /// of `produceMore()` to continue producing messages.
+            case stopProducing(CheckedContinuation<Void, Never>?)
+            /// The system is shut down.
+            case finished
+        }
+
+        /// The current state of the state machine.
+        var state = State.initial
+
+        /// The possible actions for the poll loop.
+        enum PollLoopAction {
+            /// Ask `librdkakfa` to receive new message acknowledgements at a given poll interval.
+            case pollAndSleep
+            /// Suspend the poll loop.
+            case suspendPollLoop
+            /// Shutdown the poll loop.
+            case shutdownPollLoop
+        }
+
+        /// Determines the next action to be taken in the poll loop based on the current state.
+        ///
+        /// - Returns: The next action for the poll loop.
+        func nextPollLoopAction() -> PollLoopAction {
+            switch self.state {
+            case .initial, .producing:
+                return .pollAndSleep
+            case .stopProducing:
+                // We were asked to stop producing,
+                // but the poll loop is still running.
+                // Trigger the poll loop to suspend.
+                return .suspendPollLoop
+            case .finished:
+                return .shutdownPollLoop
+            }
+        }
+
+        /// Represents the commands that can be returned by a state machine
+        /// and shall be executed by the ``KafkaBackPressurePollingSystem``.
+        enum Command {
```

Comment:
We can please stick to the `Action` naming and introduce separate actions for each method